### PR TITLE
chore: replace __proto__ by getPrototypeOf

### PR DESCRIPTION
**Pull request by kapouer**
Some users might want to run `--disable-proto=throw|delete` option for a hardened Node.
It's easy to fix and doesn't prevent `{ __proto__: null }` pattern.

### CODE REVIEW

**Comment by jfgreffier** [comment]
Should we add the `no-proto` ESLint rule in `.eslintrc.js` to enforce this ?

https://eslint.org/docs/latest/rules/no-proto

**Comment by kapouer** [comment]
Let's see how it goes

**Comment by mxschmitt** [comment]
The CI is failing: `page/page-expose-function.spec.ts:77:1 › should support throwing "null"`

**Comment by kapouer** [comment]
Oh ! fixed and rebased.

**Review by mxschmitt — APPROVED** [review]

File: packages/playwright-core/src/protocol/serializers.ts [hunk 1]
```diff
@@ -184,5 +184,6 @@ function isURL(obj: any): obj is URL {
 }
 
 function isError(obj: any): obj is Error {
-  return obj instanceof Error || obj?.__proto__?.name === 'Error' || (obj?.__proto__ && isError(obj.__proto__));
+  const proto = obj ? Object.getPrototypeOf(obj) : null;
```
Comment (yury-s) [inline packages/playwright-core/src/protocol/serializers.ts:187]: This will fail if there is a user code like `Object.getPrototypeOf = () => {};`

Comment (yury-s) [reply packages/playwright-core/src/protocol/serializers.ts:187]: Oh, this is a server side code, we don't run random js here so it should be ok.

**Review by yury-s — COMMENTED** [review]
Could you add a test for this?

File: packages/playwright-core/src/protocol/serializers.ts (hunk 1 above)
Comment (kapouer) [reply packages/playwright-core/src/protocol/serializers.ts:187]: Indeed, and `obj.__proto__` was an order of magnitude less safe than Object.getPrototypeOf :)

**Comment by kapouer** [comment]
> Could you add a test for this?

What do you mean ? this is already tested. The behavior didn't change at all ?

**Comment by kapouer** [comment]
For the record, the more modules are going to fix this, the more node is likely to get rid of __proto__ ! https://github.com/nodejs/node/issues/31951

**Comment by yury-s** [comment]
> > Could you add a test for this?
> 
> What do you mean ? this is already tested. The behavior didn't change at all ?

I mean a test that would fail before your change and pass after, this way we can ensure that this is documented and will not break in the future during refactoring of the code.

**Comment by jfgreffier** [comment]
> > > Could you add a test for this?
> > 
> > 
> > What do you mean ? this is already tested. The behavior didn't change at all ?
> 
> I mean a test that would fail before your change and pass after, this way we can ensure that this is documented and will not break in the future during refactoring of the code.

Not a test, but the `no-proto` ESLint rule that was added will guarantee that `__proto__` will not be used again

**Comment by kapouer** [comment]
Maybe `process.env.NODE_OPTIONS = (process.env.NODE_OPTIONS || "") + " --disable-proto=throw"` should be added but honestly I'm not sure where.

**Comment by kapouer** [comment]
> I'm fine with the change, but I don't get the idea behind the option above. We can't guarantee Playwright operation in the mode above, because we use third party modules and `__proto__` is a part of JavaScript. So we can land it, but it won't mean anything in the context of the above request.

`__proto__` is a *deprecated* part of js, and, while still used here and there, it is not so far away from being abandoned by modules authors.
Thanks !

**Pull request merged**